### PR TITLE
[IMP] iap, web, web_tour: improve core Dialog API

### PR DESCRIPTION
**Pull request by sdegueldre**
Previously, when trying to create a custom dialog, one would need to
extend the base Dialog class and configure it directly on the class,
using separate templates for the body and footer as necessary. This is
very unidiomatic owl code, and it is much more natural to simply extend
Component, and use Dialog at the root of the template, giving it slots
and props to configure its properties and contents.

This commit changes the dialog API to work as described and adapts code
that uses it.

### CODE REVIEW

File: addons/web/static/src/core/dialog/dialog_service.js [hunk 1]
```diff
@@ -24,38 +23,39 @@ const { EventBus } = owl;
 export const dialogService = {
     /** @returns {DialogServiceInterface} */
     start() {
-        const dialogs = {};
-        const bus = new EventBus();
+        const dialogs = reactive({});
+        window.dialogs = dialogs;
```
Comment (aab-odoo) [inline addons/web/static/src/core/dialog/dialog_service.js:27]: lol

**Review by aab-odoo — COMMENTED** [review]
Seems good. I would rename props "renderHeader" and "renderFooter" into "header" and "footer"

File: addons/iap/static/src/js/insufficient_credit_error_handler.js [hunk 2]
```diff
@@ -30,11 +31,11 @@ class InsufficientCreditDialog extends Dialog {
 
     buyCredits() {
         window.open(this.url, "_blank");
-        this.close();
+        this.props.close();
     }
 }
-InsufficientCreditDialog.bodyTemplate = "iap.redirect_to_odoo_credit";
-InsufficientCreditDialog.footerTemplate = "iap.InsufficientCreditFooter";
+InsufficientCreditDialog.components = { Dialog };
```
Comment (aab-odoo) [inline addons/iap/static/src/js/insufficient_credit_error_handler.js:37]: missing template

File: addons/iap/static/src/xml/iap_templates.xml [hunk 3]
```diff
@@ -2,23 +2,24 @@
...
 
     <!-- LAYOUT TEMPLATES -->
-    <div t-name="iap.redirect_to_odoo_credit" t-att-style="style" owl="1">
-        <t t-if="props.errorData.body">
-            <div t-raw="props.errorData.body"/>
-        </t>
-        <t t-if="!props.errorData.body">
-            <t t-if="props.errorData.message">
-                <span t-esc="props.errorData.message"/>
+    <div t-name="iap.InsufficientCreditDialog" t-att-style="style" owl="1">
```
Comment (aab-odoo) [inline addons/iap/static/src/xml/iap_templates.xml:5]: t t-name

**Review by aab-odoo — APPROVED** [review]

File: addons/web/static/src/core/dialog/dialog_service.js (hunk 1 above)
Comment (ged-odoo) [inline addons/web/static/src/core/dialog/dialog_service.js:27]: remove this

File: addons/web/static/src/webclient/user_menu/user_menu_items.xml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <templates xml:space="preserve">
 
-    <t t-name="web.UserMenu.shortcutsTable" owl="1">
-        <div>
+    <t t-name="web.UserMenu.ShortcutsDialog" owl="1">
+        <Dialog title="this.constructor.title">
```
Comment (ged-odoo) [inline addons/web/static/src/webclient/user_menu/user_menu_items.xml:5]: maybe just use directly the string `Shortcuts` here

Comment (ged-odoo) [reply addons/web/static/src/webclient/user_menu/user_menu_items.xml:5]: ah non

**Comment by ged-odoo** [comment]
robodoo r+

**Comment by sdegueldre** [comment]
@odoo/rd-security A line with a t-raw was indented. This line renders an html error message coming from our IAP servers, which I think we consider trustworthy.

**Comment by mart-e** [comment]
@robodoo override=ci/security
Reason: coming from templates on IAP side.
Safe but could be nice if we can avoid this, cf discussion on rd-fun-ia on Discord.

**Pull request closed without merging**